tools/cachegrand-benchmark-generator/src/builder.c

```diff
@@ -0,0 +1,103 @@
+/**
+ * Copyright (C) 2018-2022 Vito Castellano
+ * All rights reserved.
+ *
+ * This software may be modified and distributed under the terms
+ * of the BSD license.  See the LICENSE file for details.
+ **/
+
+#include <stdlib.h>
+#include <stdbool.h>
+
+#include "builder.h"
+
+section_t* builder_new_section_p() {
+    section_t *section;
+    section = malloc(sizeof(section_t));
+    section->subsections = malloc(sizeof(section_t));
+    section->commands = malloc(sizeof(char*));
+    section->n_subsections = 0;
+    section->n_commands = 0;
+    return section;
+}
+
+bool builder_section_append_subsection(
+        section_t *section,
+        section_t *subsections) {
+    section->subsections = realloc(
+            section->subsections,
+            (section->n_subsections+1) * sizeof(section_t));
+
+    section->subsections[section->n_subsections] = subsections;
+    if (section->subsections[section->n_subsections] != NULL) {
+        section->n_subsections = section->n_subsections+1;
+        return true;
+    }
+
+    return false;
+}
+
+bool builder_section_append_command(
+        section_t *section,
+        char *command) {
+    section->commands = realloc(
+            section->commands,
+            (section->n_commands+1) * sizeof(char*));
+
+    section->commands[section->n_commands] = command;
+    if (section->commands[section->n_commands] != NULL) {
+        section->n_commands = section->n_commands+1;
+        return true;
+    }
+
+    return false;
+}
+
+
+test_t* builder_new_test_p() {
+    test_t *test;
+    test = malloc(sizeof(test_t));
+    test->sections = (section_t**) malloc(1 * sizeof(section_t));
```

Comment:
As above, if n_sections is empty sections should be NULL especially because there is a realloc afterwards